File: clients/python/games/connect_four.py
from typing import List, Optional
from games.base import BaseGame
from framework.i18n import t
import json

ROWS, COLS = 6, 7


class ConnectFour(BaseGame):
    def __init__(self):
        self.board = [[None] * COLS for _ in range(ROWS)]
        self.players: List[str] = []
        self._turn_idx = 0
        self._over = False
        self._winner = None

    def start(self, players: List[str]) -> None:
        self.players = players
        self._turn_idx = 0
        self._over = False
        self._winner = None
        self.board = [[None] * COLS for _ in range(ROWS)]

    def current_turn(self) -> Optional[str]:
        return self.players[self._turn_idx] if self.players else None

    def validate_move(self, player_id: str, move_data: dict) -> bool:
        if self._over or player_id != self.current_turn():
            return False
        col = move_data.get('col')
        return isinstance(col, int) and 0 <= col < COLS and self.board[0][col] is None

    def apply_move(self, player_id: str, move_data: dict) -> None:
        col = move_data['col']
        for row in range(ROWS - 1, -1, -1):
            if self.board[row][col] is None:
                self.board[row][col] = player_id
                break
        if self._check_win(player_id):
            self._over = True
            self._winner = player_id
        elif all(self.board[0][c] is not None for c in range(COLS)):
            self._over = True
        else:
            self._turn_idx = 1 - self._turn_idx

    def _check_win(self, pid: str) -> bool:
        dirs = [(0, 1), (1, 0), (1, 1), (1, -1)]
        for r in range(ROWS):
            for c in range(COLS):
                if self.board[r][c] != pid:
                    continue
                for dr, dc in dirs:
                    if all(0 <= r + dr * i < ROWS and 0 <= c + dc * i < COLS and
                           self.board[r + dr * i][c + dc * i] == pid for i in range(1, 4)):
                        return True
        return False

    def is_over(self):
        return self._over, self._winner

    def render(self, perspective: Optional[str] = None) -> str:
        syms = {None: '.'}
        if self.players:
            syms[self.players[0]] = 'X'
            if len(self.players) > 1:
                syms[self.players[1]] = 'O'
        lines = [t('connect4.title')]
        for row in self.board:
            lines.append('  ' + ' '.join(syms.get(c, '?') for c in row))
        lines.append('  ' + ' '.join(str(i) for i in range(COLS)))
        return '\n'.join(lines)

    def get_state(self, perspective: Optional[str] = None) -> dict:
        return {'board': [row[:] for row in self.board],
                'turn': self.current_turn(), 'players': self.players}

    def load_state(self, data: dict, perspective: Optional[str] = None) -> None:
        if not data:
            return
        if 'players' in data:
            self.players = list(data['players'])
        if 'board' in data:
            self.board = [list(row) for row in data['board']]
        if 'turn' in data and data['turn'] in self.players:
            self._turn_idx = self.players.index(data['turn'])

    def parse_input(self, raw: str) -> Optional[dict]:
        raw = raw.strip()
        if raw.startswith('{'):
            try:
                obj = json.loads(raw)
                if isinstance(obj, dict):
                    return obj
            except ValueError:
                pass
        try:
            return {'col': int(raw.split()[0])}
        except (ValueError, IndexError):
            return None

    def get_help(self) -> List[str]:
        return [
            'Drop a piece into a column. First to connect 4 in a row wins.',
            'Move: <col>   e.g. "3"',
        ]

File: clients/node/src/games/connect_four.js
'use strict';
const { BaseGame } = require('./base');
const { t } = require('../framework/i18n');

const ROWS = 6;
const COLS = 7;

class ConnectFour extends BaseGame {
  constructor() {
    super();
    this.board = Array.from({ length: ROWS }, () => Array(COLS).fill(null));
    this.players = [];
    this._turnIdx = 0;
    this._over = false;
    this._winner = null;
  }

  start(players) {
    this.players = players;
    this._turnIdx = 0;
  }

  currentTurn() {
    return this.players[this._turnIdx] || null;
  }

  validateMove(playerId, moveData) {
    if (this._over || playerId !== this.currentTurn()) return false;
    const col = moveData.col;
    return Number.isInteger(col) && col >= 0 && col < COLS && this.board[0][col] === null;
  }

  applyMove(playerId, moveData) {
    const col = moveData.col;
    for (let row = ROWS - 1; row >= 0; row--) {
      if (this.board[row][col] === null) {
        this.board[row][col] = playerId;
        break;
      }
    }
    if (this._checkWin(playerId)) {
      this._over = true;
      this._winner = playerId;
    } else if (this.board[0].every(c => c !== null)) {
      this._over = true;
      this._winner = null;
    } else {
      this._turnIdx = 1 - this._turnIdx;
    }
  }

  _checkWin(pid) {
    const b = this.board;
    const dirs = [[0, 1], [1, 0], [1, 1], [1, -1]];
    for (let r = 0; r < ROWS; r++) {
      for (let c = 0; c < COLS; c++) {
        if (b[r][c] !== pid) continue;
        for (const [dr, dc] of dirs) {
          let win = true;
          for (let i = 1; i < 4; i++) {
            const nr = r + dr * i, nc = c + dc * i;
            if (nr < 0 || nr >= ROWS || nc < 0 || nc >= COLS || b[nr][nc] !== pid) {
              win = false; break;
            }
          }
          if (win) return true;
        }
      }
    }
    return false;
  }

  isOver() {
    return [this._over, this._winner];
  }

  render(perspective) {
    const syms = { null: '.' };
    if (this.players[0]) syms[this.players[0]] = 'X';
    if (this.players[1]) syms[this.players[1]] = 'O';
    const lines = [t('connect4.title')];
    for (const row of this.board) {
      lines.push('  ' + row.map(c => syms[c] || '?').join(' '));
    }
    lines.push('  ' + Array.from({ length: COLS }, (_, i) => i).join(' '));
    return lines.join('\n');
  }

  getState(perspective) {
    return { board: this.board, turn: this.currentTurn(), players: this.players };
  }

  loadState(data) {
    if (!data) return;
    if (data.players) this.players = data.players;
    if (data.board) this.board = data.board.map(row => [...row]);
    if (data.turn != null) {
      const idx = this.players.indexOf(data.turn);
      this._turnIdx = idx >= 0 ? idx : 0;
    }
  }

  parseInput(raw) {
    const trimmed = raw.trim();
    if (trimmed.startsWith('{')) {
      try { const obj = JSON.parse(trimmed); if (obj && typeof obj === 'object') return obj; } catch (_) {}
    }
    const n = Number(trimmed.split(/\s+/)[0]);
    return isNaN(n) ? null : { col: n };
  }

  getHelp() {
    return [
      'Drop a piece into a column. First to connect 4 in a row wins.',
      'Move: <col>   e.g. "3"',
    ];
  }
}

module.exports = { ConnectFour };
